### ISSUE TITLE
Account for minimum size of windows 

### ISSUE THREAD

**Issue by blahsd** [koekeishiya/chunkwm#324]
Some windows won't resize pasta certain point. Therefore when tiling they will get the correct position but keep a size that can be larger than their leaf, messing up the whole display. I'm attaching a screenshot that displays the issue. 
<img width="1680" alt="screenshot 2018-01-27 13 44 52" src="https://user-images.githubusercontent.com/7347353/35472043-46a3e1d4-0368-11e8-9989-dde5b6ccb1e9.png">

In this picture you can see that the iTunes window won't scale down to the size its supposed to be. Here it's not a big deal but when the desktop is fuller it can become a real nuisance, and windows can start overlapping. At the point in which some windows are covering other windows, this (at least for me) largely defeats the purpose of using a window manager, which I rely on to have all windows nicely separated and visible. 

There are two possible solutions:

1. Forcing the window to scale down. This off would be preferable but I'm not sure it's possible or within the goals of this tool to enable forced scale-down of unresizable windows. 
2. Expanding the region automatically in case that a window can't fit in its leaf. This is less preferable because it would could default to unexpected window sizes, but at least you don't get overlaps.

**Comment by koekeishiya** [comment]
I very much agree that this is an annoying problem. Unfortunately this is a hard problem to solve within the constraints we are given.

> 1. Forcing the window to scale down. This off would be preferable but I'm not sure it's possible or within the goals of this tool to enable forced scale-down of unresizable windows.

This sounds like the most appropriate solution, but I do not think there are any APIs that we can use to do this. Lets say we did find such an API, I assume we would then somehow determine what the scale should be, given the minimum-size and the target-size that we want.

> 2. Expanding the region automatically in case that a window can't fit in its leaf. This is less preferable because it would could default to unexpected window sizes, but at least you don't get overlaps.

This seems to be the most realistic way of dealing with this situation, however, this is not something that can be applied universally. Consider the case where two (or more) windows with fairly large minimum-size constraints are attempted to be tiled next to eachother. How do we resolve that conflict and what is the expected behaviour?

What I personally do to deal with this, is that I sort of know by now the amount of space that is optimal for the different applications that I use, and I simply open them in a desktop where that is appropriate.

I do encourage people to discuss this further however; if there is a potential solution to be implemented, that would be great.

**Comment by dominiklohmann** [comment]
> Consider the case where two (or more) windows with fairly large minimum-size constraints are attempted to be tiled next to eachother. How do we resolve that conflict and what is the expected behaviour?

Assuming there are just two windows: Display the focused window fully, move the unfocused window such that the overlapping area is to the outside of the screen to always keep the gap intact.

I do not know how to properly handle this with 3+ windows, as it could be impossible to keep the gap intact.

> The other case would be that increasing the region for the window with a minimum size, could trigger the other window to hit its minimum-size (but it was fine before we forced the region to change) and we would just endlessly loop.

In that case, don't change the default tile ratio of either leaf.

**Comment by koekeishiya** [comment]
@dominiklohmann 

I think what you suggested seems reasonable, however I know of no way to actually retrieve the minimum / maximum size a window can have, so it would be hard to identify when to not change things, as mentioned in the latter case.

If anybody knows how we can query the size constraints of a window, that would be huge.

**Comment by blahsd** [comment]
Let me say that the amount of attention you dedicate to this tool and to your user base is outright flattering. Your responses are lighting fast. 

> This seems to be the most realistic way of dealing with this situation, however, this is not something that can be applied universally. Consider the case where two (or more) windows with fairly large minimum-size constraints are attempted to be tiled next to eachother. How do we resolve that conflict and what is the expected behaviour?

My preferred solution (but this is very personal) would be that a window that doesn't fit in its assigned leaf gets sent to a different desktop. 

> If anybody knows how we can query the size constraints of a window, that would be huge.

I don't think you can query directly the size constraints. I think the most realistic approach would be getting the size of the window after it's been tiled and comparing it to the size of the leaf. I think you can query that – I know no C++ specifically, but according to the internet you should be able to get through: something like:

```
RECT rect;
if(GetWindowRect(hwnd, &rect))
{
  int width = rect.right - rect.left;
  int height = rect.bottom - rect.top;
}
```

So, summing up, ideally:

- New window opens and gets tiled like any other
- Verify if the window's size is equal or lower than the leaf's size. If so, do nothing.
- If not, move the window to the next desktop. 

Hope this helps. I wish I could provide pull requests instead of wishful thinking.

**Comment by koekeishiya** [comment]
Unfortunately `var contentMinSize: NSSize { get set }` is only accessible through the `NSWindow` handle, which is only accessible internally to the application that owns the window.

> Otherwise, perhaps more easily, you could query the size of the window and compare it to the leaf's size after it's been tiled to check whether it's respecting the constraints.

We probably have to do something like this indeed.

Regarding moving the window to some other desktop, in theory this seems nice, but what if there are some monocle desktops, and some floating desktops. Which desktop is the window supposed to be moved to. I feel like there are just so many different options when we try to automate this task.
Having config settings for all of this would be crazy (imo), and so I gotta wonder if it is something that we truly should automate at all.

I need to think through this properly before any attempt is made at least.

**Comment by kerryj89** [comment]
Maybe it was already suggested but I wouldn't mind having an option to align the offending window to the monitor edges so that it's always visible on screen, albeit overlapping other windows.  Ugly but more functional than having part of it inaccessible without finagling it.
